pkg/daemon/pinned_image_set.go

```diff
@@ -0,0 +1,1273 @@
+
+	p.mcpLister = mcpInformer.Lister()
+	p.mcpSynced = mcpInformer.Informer().HasSynced
+
+	p.cache = lru.New(256)
+
+	return p
+}
+
+type imageInfo struct {
+	Name   string
+	Size   int64
+	Pulled bool
+}
+
+func (p *PinnedImageSetManager) sync(key string) error {
+	klog.V(4).Infof("Syncing PinnedImageSet for pool: %s", key)
+	node, err := p.nodeLister.Get(p.nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to get node %q: %w", p.nodeName, err)
+	}
+
+	pools, _, err := helpers.GetPoolsForNode(p.mcpLister, node)
+	if err != nil {
+		return err
+	}
+
+	if len(pools) == 0 {
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), p.prefetchTimeout)
+	// cancel any currently running tasks in the worker pool
+	p.resetWorkload(cancel)
+
+	if err := p.syncMachineConfigPools(ctx, pools); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			ctxErr := fmt.Errorf("requeue: prefetching images incomplete after: %v", p.prefetchTimeout)
+			p.updateStatusError(pools, ctxErr)
+			klog.Error(ctxErr)
+			return ctxErr
+		}
+		p.updateStatusError(pools, err)
+		return err
+	}
+
+	p.updateStatusProgressingComplete(pools, "All pinned image sets complete")
+	return nil
+}
+
+func (p *PinnedImageSetManager) syncMachineConfigPools(ctx context.Context, pools []*mcfgv1.MachineConfigPool) error {
+	images := make([]mcfgv1alpha1.PinnedImageRef, 0, 100)
+	for _, pool := range pools {
+		validTarget, err := validateTargetPoolForNode(p.nodeName, p.nodeLister, pool)
+		if err != nil {
+			klog.Errorf("failed to validate pool %q: %v", pool.Name, err)
+			return err
+		}
+		if !validTarget {
+			continue
+		}
+
+		if err := p.syncMachineConfigPool(ctx, pool); err != nil {
+			return err
+		}
+
+		// collect all unique images from all pools
+		for _, image := range pool.Spec.PinnedImageSets {
+			imageSet, err := p.imageSetLister.Get(image.Name)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					klog.Warningf("PinnedImageSet %q not found", image.Name)
+					continue
+				}
+				return fmt.Errorf("failed to get PinnedImageSet %q: %w", image.Name, err)
+			}
+			images = append(images, imageSet.Spec.PinnedImages...)
+		}
+	}
+
+	// verify all images are pulled and available if not clear the cache and requeue
+	for _, image := range images {
+		exists, err := p.criClient.ImageStatus(ctx, image.Name)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			p.cache.Clear()
+			return errFailedToPullImage
+		}
+	}
+
+	//  write config and reload crio last to allow a window for kubelet to gc images in emergency
+	if err := p.ensureCrioPinnedImagesConfigFile(images); err != nil {
+		klog.Errorf("failed to write crio config file: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// validateTargetPoolForNode checks if the node is a target for the given pool.
+func validateTargetPoolForNode(nodeName string, nodeLister corev1lister.NodeLister, pool *mcfgv1.MachineConfigPool) (bool, error) {
```

Comment:
So I was trying to understand this function's usage (since normally you're only allowed to be in 1 pool for every node). Could you help me check if my understanding is correct:

I guess this came from `GetPoolsForNode` which looks like we were using for metrics? It seems to allow combinations of `master, custom, worker` and `custom, worker` which I guess this is being used differently in the sense that we want to check which pinnedimagesets should attach to what nodes.

i.e. ultimately, if you set a PIS for a custom pool, it will only apply to custom pools
if you set a PIS for worker, this will actually apply it to worker+custom so long as you still have the worker role on the node
if you set a PIS for hybrid master/worker nodes (i.e. a node with both master and worker labels) you get master
and for hybrid master/worker/custom cluster nodes you'll get PIS targetted for both master and worker?

Is that understanding correct? I'm not actually sure if that logic is... entirely sound, now I look at it

Reply:
right given a pool is this node part of that pool.

Reply:
I also saw it in the node controller is this more inline with your expectations?

https://github.com/openshift/machine-config-operator/blob/204767253e30608b5b7fd70ad1ace02ba1d64b46/pkg/controller/node/node_controller.go#L944-L971

Reply:
the important difference is that it focuses on the primary pool for the node.

Reply:
OK good point I am going to write a test to validate my theory

Reply:
To answer the question the approach taken here is to emulate MC.

In the case of a custom pool given the example

```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: infra
spec:
  machineConfigSelector:
    matchExpressions:
      - {key: machineconfiguration.openshift.io/role, operator: In, values: [worker,infra]}
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/infra: ""
```
In the case where we have a worker define PinnedImageSet all of the nodes labeled worker and or infra.

Reply:
Hmm, so with the updated method, I think we still have some diffs against MachineConfig behaviour. (If I'm understanding correctly, please correct if my logic is wrong, it looks like `nodes, err := nodeLister.List(selector)` will just blindly select with regards to labels on the node) With PinnedImageSets:

 - a node that only has `infra` label will only get `infra` PinnedImageSets
 - a node that is `master, worker` will get both `master, worker` PinnedImageSets

Whereas for machineconfigs:
 - a node that only has `infra` label will get both infra and worker MachineConfigs
 - a node that is `master, worker` will only get master MachineConfigs

If we're fine with that behaviour diff (basically labels are what determines PinnedImageSets and not really a pool match)

Reply:
sorry I had not yet pushed

Reply:
I dropped that whole method as the decision for what pool gets a set is defined by the pinned image set controller.

Reply:
To be clear I don't want a diff with MC so there is less possible confusion. Today this feature is defined by pools so it should align with existing MCP principles.

Reply:
see https://github.com/openshift/machine-config-operator/pull/4303/files#diff-bda0abaca253395b277975ae62c03e13a4b9cae1fafcf32da282b859a0e7825bR55

Reply:
see https://github.com/openshift/machine-config-operator/pull/4303/files#diff-bda0abaca253395b277975ae62c03e13a4b9cae1fafcf32da282b859a0e7825bR55